### PR TITLE
Speed improvements for patch embeddings 

### PR DESCRIPTION
**Pull request by picsalex**
### Summary

On large datasets (110k+ images, 2M+ bboxes), `yolo-scout` would crash or take 40+ hours in the embeddings and quality-metrics steps. This PR fixes four independent, compounding issues and adds a real (if modest) speed improvement, validated end-to-end on an 11.7k-image / 1.7M-patch production dataset.

### Changes

**1. Unbounded memory growth in patch processing** (`embeddings/preprocessing.py`, `embeddings/computer.py`, `visualization/quality.py`)

Patch-crop extraction and quality-metric computation each materialized *every* crop for the *entire* dataset into memory before doing anything with it (`list(pool.imap(...))`). At 2M crops this could reach 100+ GB resident and crash the embedded MongoDB. `extract_all_patch_crops` → `iter_patch_crops`, a generator that streams `(sample_id, crops)` per sample so callers can process-and-discard in bounded batches. `quality.py`'s own duplicate copy of the same anti-pattern was replaced with the same generator.

**2. Worker/thread oversubscription** (`embeddings/computer.py`, `embeddings/preprocessing.py`, `cli.py`)

Several libraries independently default to using every CPU core, and on high-core-count machines they stack:
- FiftyOne's image-embeddings `DataLoader` (`cpu_count() // 2` workers, each opening its own connection to the embedded mongod) — capped to 8, since a high worker count was crashing mongod.
- PyTorch's own intra-op thread pool — capped to 4 via `torch.set_num_threads()`.
- `numba`'s thread pool (used internally by UMAP/pynndescent) — defaults to full CPU count and, unlike torch, can't be shrunk at runtime once launched. Fixed via `NUMBA_NUM_THREADS` env var set as the first line of `cli.py`, before numba can be imported anywhere.
- OpenCV's internal per-call threading — capped to 1 per worker process via a `Pool` initializer, since the pool's ~150+ worker *processes* already provide all the parallelism needed.

**3. UMAP segfault at scale** (`embeddings/computer.py`)

FiftyOne's UMAP wrapper hardcodes `init="spectral"`, which at >~1.7M points uses SciPy's ARPACK eigensolver and segfaults (native crash, no Python traceback). Patch-embeddings visualization now runs UMAP directly with `init="pca"` (safe, deterministic, avoids the ARPACK path entirely) and hands FiftyOne the result via `points=` / `method="manual"`, using FiftyOne Brain's own `get_embeddings()` to guarantee correct sample/label association.

**4. Quality metrics computed in the wrong place** (`visualization/quality.py`, `embeddings/preprocessing.py`)

Per-patch quality metrics (`cv2.Laplacian`, `cv2.calcHist`, etc.) were computed serially in the main process after crop extraction, leaving the large extraction pool idle for that portion. `iter_patch_crops` now accepts an optional `worker_func`, and quality metrics are computed *inside* the same pool workers, alongside extraction — measured ~5% faster.

### Net effect (production dataset: 11.7k images, 1.7M patches)

- Patch embeddings: crash / never completed → ~45 min
- Quality metrics: crash (211GB RES) / 40+ hr projection → ~2h51m
- No more segfaults, no more OOM, fully deterministic UMAP output

### CODE REVIEW

File: yolo_scout/visualization/quality.py [hunk 1]
```diff
@@ -85,36 +106,22 @@ def get_patches(sample):
...
         patches = get_patches(sample)
-        for patch, crop in zip(patches, crops):
-            patch_gray = cv2.cvtColor(crop, cv2.COLOR_RGB2GRAY)
-            patch["blurriness"] = _blurriness(patch_gray)
-            patch["brightness"] = _brightness(patch_gray)
-            patch["aspect_ratio"] = _aspect_ratio(patch_gray)
-            patch["entropy"] = _entropy(patch_gray)
+        for patch, metrics in zip(patches, metrics_list):
+            patch["blurriness"] = metrics["blurriness"]
+            patch["brightness"] = metrics["brightness"]
```
[the remark this answers is not in the document]
Comment (picsalex) [reply yolo_scout/visualization/quality.py]: @copilot should I add an empty patch (gray-like) or just handle None values instead so we don't have skipped patches?

[the remark this answers is not in the document]
Comment (picsalex) [reply yolo_scout/visualization/quality.py]: Skipping Copilot's review and re-implementing it with Claude

**Review by onuralpszr — APPROVED** [review]
